### ISSUE TITLE
Status of ImagickMagick 7

### ISSUE THREAD

**Issue by remicollet** [mkoppanen/imagick#149]
Howdy,

Trying to build against ImageMagick 7.0.1, I encounter some issue with missing functions

With 3.4.2:

```
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/shim_im6_to_im7.c:319:11: warning: implicit declaration of function 'MagickFilterImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/shim_im6_to_im7.c:567:11: warning: implicit declaration of function 'MagickOrderedPosterizeImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagickdraw_class.c:1722:12: warning: implicit declaration of function 'DrawGetStrokeOpacity' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:299:10: warning: implicit declaration of function 'MagickGetImageMatte' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:537:11: warning: implicit declaration of function 'MagickRandomThresholdImageChannel' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:753:16: warning: implicit declaration of function 'MagickTransformImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:1036:16: warning: implicit declaration of function 'MagickGetImageInterpolateMethod' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:1059:11: warning: implicit declaration of function 'MagickSetImageInterpolateMethod' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:1240:12: warning: implicit declaration of function 'MagickPaintFloodfillImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:2200:6: warning: implicit declaration of function 'MagickGetImageChannelRange' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:2259:11: warning: implicit declaration of function 'MagickGetImageChannelDistortion' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:2756:11: warning: implicit declaration of function 'MagickGetImageChannelKurtosis' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:4260:11: warning: implicit declaration of function 'MagickColorFloodfillImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:4686:11: warning: implicit declaration of function 'MagickMapImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:4725:11: warning: implicit declaration of function 'MagickMatteFloodfillImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:4843:11: warning: implicit declaration of function 'MagickPaintOpaqueImageChannel' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:4924:11: warning: implicit declaration of function 'MagickPaintTransparentImage' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:5155:11: warning: implicit declaration of function 'MagickSeparateImageChannel' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:5221:11: warning: implicit declaration of function 'MagickSetImageBias' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:5358:11: warning: implicit declaration of function 'MagickSetImageChannelDepth' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:6565:18: warning: implicit declaration of function 'MagickGetImageChannelDepth' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:6626:11: warning: implicit declaration of function 'MagickGetImageChannelExtrema' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:6660:11: warning: implicit declaration of function 'MagickGetImageChannelMean' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:6963:11: warning: implicit declaration of function 'MagickGetImageExtrema' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:7198:11: warning: implicit declaration of function 'MagickGetImageMatteColor' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:8071:11: warning: implicit declaration of function 'MagickGetImageIndex' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:8186:58: warning: implicit declaration of function 'MagickGetImageSize' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:8658:11: warning: implicit declaration of function 'MagickSetImageAttribute' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:9441:11: warning: implicit declaration of function 'MagickSetImageMatteColor' [-Wimplicit-function-declaration]
/dev/shm/BUILD/php-pecl-imagick-3.4.2/ZTS/imagick_class.c:12540:12: warning: implicit declaration of function 'MagickFilterImage' [-Wimplicit-function-declaration]
```

And of course, extension can't be loaded

```
 PHP Warning:  PHP Startup: Unable to load dynamic library 'modules/imagick.so' - modules/imagick.so: undefined symbol: MagickGetImageChannelDistortion in Unknown on line 0
```

(same issue with git master)

**Comment by Danack** [comment]
I've turned on this error detection and fixed the errors. For some of them e.g. [MagickSetImageInterpolateMethod](https://github.com/ImageMagick/ImageMagick/issues/181) the function appears to be missing entirely from ImageMagick

> And of course, extension can't be loaded

I'm currently not building it statically - so actually I don't get these errors. I probably ought to do that.

**Comment by remicollet** [comment]
> I'm currently not building it statically - so actually I don't get these errors. I probably ought to do that.

Not I build it shared ;)
But the default packaged PHP version is patch to dlopen extension using  RTLD_NOW (instead of RTLD_LAZY), which mean all symbols must be resolved at load time (which seems far better than segfaulting later at runtime)

**Comment by Danack** [comment]
@remicollet Do you know anyone who has the time/energy who could submit a patch to add autoconf 'probe' program for detecting whether `MagickSetImageInterpolateMethod` is present in the ImageMagick library or not?

I tried to do this before, and decided that I would rather give up using computers than have to spend a significant amount of time wrestling with autoconf.

> far better than segfaulting later at runtime

I prefer the term _dynamically stable_.

**Comment by remicollet** [comment]
See PR #150 

Notice, with this, build against 7.0.1-0 works, only 2 failed tests.

```
=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Test Imagick, forwardFourierTransformImage [tests/073_Imagick_forwardFourierTransformImage_basic.phpt]
Test Imagick, forwardFourierTransformImage [tests/086_Imagick_forwardFourierTransformImage_basic.phpt]
=====================================================================

```

**Comment by Danack** [comment]
Thanks. 

I think everything is just about working now. If you agree, I'll tag another release.

**Comment by remicollet** [comment]
Yes everything seems ok now.

Test suite pass with IM 7.0.1-0 and PHP 5.4.45, 5.5.35, 7.6.21 and 7.0.6
Only some build warnings

**Comment by remicollet** [comment]
BTW, in your opinion, which version of IM should I use for the extension package, as both are available ?

It seems we loose lot of methods using 7.x, see http://paste.fedoraproject.org/361973/62253691/

**Comment by Danack** [comment]
That is a question that has a answer with complex parts:

i) I think we ought to ask the ImageMagick guys what the selling points of IM7 are, so that we can point users to them if they ask.

ii) A lot of the functions that are no longer available are bogus and shouldn't have been used for ages. Basically the IM guys have kept the concept of 'opacity' and 'matte' around for BC reasons, even though they moved to a pure 'alpha' about 5 years ago, so a lot of the removed functions are good to be gone.

iii) I'm pretty sure getImageAttribute was also deprecated. The explanation of what to replace it was confusing to me.

iv) I'm not sure that setimagebias has ever worked reliably - https://github.com/ImageMagick/ImageMagick/issues/125 For ImageMagick 7 the whole idea of 'bias' has been removed. Instead users are now able to do any of the functions that involved 'bias' themselves using the morphology function http://php.net/manual/en/imagick.morphology.php and a user defined kernel.

I would show you an example at http://phpimagick.com/Imagick/morphology?morphologyType=1 but it appears I managed to break this on my site.

v) I hadn't realised/taken note of getimageextrema and getimagechannelextrema were gone. I'm not sure what should be used instead.

vi) setimageindex and getimageindex are replaced with setIteratorIndex and getIteratorIndex. However.....the iterating over images does not work as people expect - in particular, the 'individual images' are not 'detached' when iterating, so for example when iterating over an image with three frames:

```
$imagick = new Imagick("SomeGifWithThreeFrames.gif");
$blankCanvas = new Imagick();
foreach ($imagick as $imagickFrame) {
     $blankCanvas->addImage($imagickFrame);
}

```

Most people would expect `$blankCanvas` to now also contain 3 frames.....but actually it contains 9 image frames.....I think this makes it be classified as a false affordance, and so a bad abstraction to have.

This is this list of things I'm thinking of tidying up for version 4: https://github.com/mkoppanen/imagick/labels/V4 including the [iterator shenanigans](https://github.com/mkoppanen/imagick/issues/122)

TL:DR - I don't know.

I can't see an awesome reason to upgrade to IM7 straight away, but it should be done at some point. It might be worth holding onto version 6.x until I refactor Imagick to cover the things that ought to be cleaned up for version 4 of imagick.

**Comment by remicollet** [comment]
Thanks a lot for explanations.

Perhaps lost methods could be marked as "deprecated" with IM 6 ?
So people won't be surprized ;)

**Comment by Danack** [comment]
Going to close this as it doesn't seem to contain any actionable items that aren't covered elsewhere.
